### PR TITLE
ROX-20482: Enable Go mod prefetching

### PR DESCRIPTION
**Pull request by msugakov**
Without hermetic builds.

The problem was due to stale image hash of `quay.io/redhat-appstudio-tekton-catalog/task-prefetch-dependencies:0.1` and insufficient workspace volume.

----

How I validated the change - by comparing pipeline run logs.

A build with prefetching enabled. `prefetch-dependencies` takes 1m40s, `build-container` takes 5m31s.
Interestingly, the build system injects `. /cachi2/cachi2.env &&` into every `RUN` statement in the dockerfile.
I.e. the original
```
RUN mkdir -p image/bin
```
becomes
```
RUN . /cachi2/cachi2.env &&     mkdir -p image/bin
```
That's how, I believe, they make sure all appropriate commands (`go ...`) find cached packages.

A build without prefetching. `prefetch-dependencies` obviously gets skipped. `build-container` takes 5m29s.
No speedup in build and overall slowdown, unfortunately.
Plenty of logs like
```
go: downloading github.com/spf13/cobra v1.7.0
go: downloading github.com/spf13/pflag v1.0.6-0.20210604193023-d5e0c0615ace
go: downloading github.com/pkg/errors v0.9.1
go: downloading golang.org/x/net v0.17.0
```
which were absent in the run with prefetch.

Question why prefetch doesn't make build faster https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1698838318724849 got the answer that Cachi2 at the moment is not for speeding up builds.

### CODE REVIEW

**Comment by tommartensen** [comment]
> @tommartensen the message about changing when: is https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1698951790776289

TLDR: Don't worry, Enterprise Contract will check task conditionals for the important ones.

**Review by tommartensen — CHANGES_REQUESTED** [review]
We have the same problem as [this](https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1698949423506559). 

```
$ docker run  quay.io/redhat-user-workloads/rh-acs-tenant/acs/roxctl:on-pr-a5e807febd4529ff2aef6b48d4943161aa7aa636 version
Unable to find image 'quay.io/redhat-user-workloads/rh-acs-tenant/acs/roxctl:on-pr-a5e807febd4529ff2aef6b48d4943161aa7aa636' locally
...
Digest: sha256:17bc86ca44dd8dc96ffec8f7a67704827c1112b255105b4182222a3cf4d174ea
Status: Downloaded newer image for quay.io/redhat-user-workloads/rh-acs-tenant/acs/roxctl:on-pr-a5e807febd4529ff2aef6b48d4943161aa7aa636
...
4.3.x-61-ga5e807febd-dirty
```

because of the modified Dockerfile.

**Comment by msugakov** [comment]
Regarding detection whether we're in the context of RHTAP v.s. building locally. There are no interesting environment variables set by Cachi2.

```
[1/2] STEP 4/8: RUN . /cachi2/cachi2.env &&     declare -xp
declare -x GOCACHE="/cachi2/output/deps/gomod"
declare -x GOFLAGS="-mod=vendor"
declare -x GOMAXPROCS="8"
declare -x GOMODCACHE="/cachi2/output/deps/gomod/pkg/mod"
declare -x GOPATH="/cachi2/output/deps/gomod"
declare -x HOME="/root"
declare -x HOSTNAME="80f3df7375b0"
declare -x OLDPWD
declare -x PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
declare -x PWD="/go/src/github.com/stackrox/rox/app"
declare -x SHLVL="1"
declare -x SUMMARY="RHEL8 based Go builder image for OpenShift ART"
declare -x VERSION="1.20"
declare -x container="oci"
```

Looks like we need to check presence of `/cachi2/cachi2.env` which isn't very reliable should they decide to move the file, or just fail unconditionally.

**Comment by msugakov** [comment]
That's how the build failure log looks when the modified files check fails.
```
step-build
Prefetched content will be made available
[1/2] STEP 1/7: FROM brew.registry.redhat.io/rh-osbs/openshift-golang-builder:rhel_8_1.20 AS builder
Trying to pull brew.registry.redhat.io/rh-osbs/openshift-golang-builder:rhel_8_1.20...
Getting image source signatures
Copying blob sha256:f1d54b6c7b464a85ce61769d6d4fb1a07eb364a5d4d6fccb6ee53a69a8d2f71d
Copying blob sha256:f70d60810c69edad990aaf0977a87c6d2bcc9cd52904fa6825f08507a9b6e7bc
Copying blob sha256:545277d800059b32cf03377a9301094e9ac8aa4bb42d809766d7355ca9aa8652
Copying config sha256:5d25d1e468630dde1b398572fc1b7c42941c0ca28250a38589487820b440ecd1
Writing manifest to image destination
[1/2] STEP 2/7: WORKDIR /go/src/github.com/stackrox/rox/app
[1/2] STEP 3/7: COPY . .
[1/2] STEP 4/7: RUN . /cachi2/cachi2.env &&     echo "Checking that files in git are not modified." &&     echo "If the following command fails, it will also print modified files." &&     echo "You need to figure the reason and prevent that because otherwise the build results will be inconsistent." &&     git status --porcelain | { ! grep '.' >&2 ; }
Checking that files in git are not modified.
If the following command fails, it will also print modified files.
You need to figure the reason and prevent that because otherwise the build results will be inconsistent.
 M image/roxctl/rhtap.Dockerfile
subprocess exited with status 1
subprocess exited with status 1
Error: building at STEP "RUN . /cachi2/cachi2.env &&     echo "Checking that files in git are not modified." &&     echo "If the following command fails, it will also print modified files." &&     echo "You need to figure the reason and prevent that because otherwise the build results will be inconsistent." &&     git status --porcelain | { ! grep '.' >&2 ; }": exit status 1

step-sbom-syft-generate
2023/11/06 17:33:30 Skipping step because a previous step failed

step-analyse-dependencies-java-sbom
2023/11/06 17:33:31 Skipping step because a previous step failed

step-merge-syft-sboms
2023/11/06 17:33:31 Skipping step because a previous step failed

step-merge-cachi2-sbom
2023/11/06 17:33:31 Skipping step because a previous step failed

step-create-purl-sbom
2023/11/06 17:33:32 Skipping step because a previous step failed

step-inject-sbom-and-push
2023/11/06 17:33:32 Skipping step because a previous step failed

step-upload-sbom
2023/11/06 17:33:33 Skipping step because a previous step failed
```

**Comment by msugakov** [comment]
`roxctl` is not dirty any more.

```
$ docker run --rm -it quay.io/redhat-user-workloads/rh-acs-tenant/acs/roxctl:on-pr-09e5188ab9e709f27bdb9d4125678348606ccd01 version
[...]
4.3.x-63-g09e5188ab9
```

File: image/roxctl/rhtap.Dockerfile [hunk 1]
```diff
@@ -12,6 +12,17 @@ WORKDIR /go/src/github.com/stackrox/rox/app
 
 COPY . .
 
+# When dependencies prefetch is on, RHTAP modifies this Dockerfile which would fail the following check.
+# This should stop happening after https://issues.redhat.com/browse/STONEBLD-1847
+# Since changes to the Dockerfile aren't needed (in its copy inside the build context), we revert changes.
+# This is only done when in RHTAP which is checked by the presence of cachi2.env file in absence of anything better.
+RUN if [[ -f /cachi2/cachi2.env ]]; then git restore image/roxctl/rhtap.Dockerfile; fi
+
+RUN echo "Checking that files in git are not modified." && \
```
Comment (tommartensen) [inline image/roxctl/rhtap.Dockerfile:21]: Have you tried combining this in one layer (ie one RUN)? 
Since we likely have to repeat this workaround for at least `main` image, it might make sense to preemptively put it in a script.

Comment (msugakov) [reply image/roxctl/rhtap.Dockerfile:21]: > Have you tried combining this in one layer (ie one RUN)?

Reducing number of `RUN`s does not win us much in the `builder` stage because it's not the one users pull. Unless you have in mind benefit other than reducing the number of pulled layers?

I'm creating a script but have difficulties finding the optimal place for it in the repo. Do you have any suggestions?

Comment (tommartensen) [reply image/roxctl/rhtap.Dockerfile:21]: > Unless you have in mind some benefit other than reducing the number of pulled layers?
You're right, the number of builder stages should not matter. 

> finding the optimal place for it in the repo
`scripts/ci/rhtap`, `.rhtap`, `.tekton/scripts` might be options

**Comment by msugakov** [comment]
After the recent change (removed `when:` altogether).

When `prefetch-input` is empty, task output is the following
```
roxctl-on-pull-request-m5pph-prefetch-dependencies
[prepare] 2023/11/07 14:46:43 Entrypoint initialization

[place-scripts] 2023/11/07 14:46:44 Decoded script /tekton/scripts/script-0-lqmft

[prefetch-dependencies] Build will be executed with network isolation, but no content was configured to be prefetched.

%!s(<nil>)
```

`build-container` downloads modules. EC-no-hermetic passed (with 3 warnings).

When `prefetch-input` is set, task output is
```
roxctl-on-pull-request-28lsd-prefetch-dependencies

step-prefetch-dependencies
2023-11-07 15:32:40,650 INFO Go version: go version go1.19.10 linux/amd64
2023-11-07 15:32:40,655 INFO Fetching the gomod dependencies at subpath .
2023-11-07 15:32:40,656 INFO Fetching the gomod dependencies at the "." directory
2023-11-07 15:32:41,047 INFO Downloading the gomod dependencies
2023-11-07 15:33:41,944 INFO Retrieving the list of packages
2023-11-07 15:33:51,811 INFO All dependencies fetched successfully \o/
```

No modules downloading during `build-container`. EC-no-hermetic passed (with 4 warnings).

Also,
```
$ docker run --rm -it quay.io/redhat-user-workloads/rh-acs-tenant/acs/roxctl:on-pr-27ca08436ec81b5d58a45b20559a24a2849a4547 version
4.3.x-65-g27ca08436e
```

File: image/roxctl/rhtap.Dockerfile (hunk 1 above)
Comment (msugakov) [reply image/roxctl/rhtap.Dockerfile]: Used `scripts/rhtap/fail-build-if-git-is-dirty.sh`.
Did not use `scripts/ci` because files there looked very much intended for CI. This file here is executed in Dockerfile which will be built not necessarily in CI (at least we intend to).

File: scripts/rhtap/fail-build-if-git-is-dirty.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    git restore image/roxctl/rhtap.Dockerfile
+fi
+
+# Next, make sure no other things that make it `-dirty` slipped through. If they did, fail the build.
+
+echo "Checking that files in git repo are not modified."
+echo "If this command fails, you should see the list of modified files below."
+echo "You need to find the reason and prevent it because otherwise the build results will be inconsistent."
+echo ""
+git status --porcelain | { ! { grep '.' >&2 && echo "ERROR: Modifies files found." >&2; } ; }
```
Comment (tommartensen) [inline scripts/rhtap/fail-build-if-git-is-dirty.sh:28]:
```suggestion
git status --porcelain | { ! { grep '.' >&2 && echo "ERROR: Modified files found." >&2; } ; }
```

small typo, plus I think the style-check is complaining about the `!` without condition.

Comment (msugakov) [reply scripts/rhtap/fail-build-if-git-is-dirty.sh]: Done.

**Review by tommartensen — APPROVED** [review]
LGTM
